{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Identifying gene expression modules using an ICA matrix decompositon. \n",
    "\n",
    "#### Motivation \n",
    "It was recently reported that there are approximately 11m deaths due to sepsis annually (Rudd et al. 2020). Most of these deaths occur in the ICU, where patients succumb despite intensive care aimed to prevent organ dysfunction and failure (Sakr et al. 2018). It is well known that exaggerated innate immune responses, notably a massive cytokine storm-driven hyper-inflammation, result in early mortality in hospital patients with sepsis (Rittirsch et al. 2008; Hotchkiss et al. 2013a). At the same time, dysregulation of compensatory anti-inflammatory mechanisms and a global down-regulation of various cytokines (i.e., immunosuppression) occur, also contributing to mortality if these responses dominate (Boomer et al. 2011). However, the interplay of these mechanisms is not well understood, and importantly, the specific molecules dysregulated are unknown. Thus, transcriptomic studies that characterize the global gene expression profiles of patients who succumb to sepsis are lacking. \n",
    "\n",
    "The typical analysis performed on gene expression datasets is differential expression (DE) analysis, where some generalized linear model (implemented in packages like DESeq2, limma, edgeR) is used to determine the association between the expression of one gene with a phenotype of interest. Accordingly, genes (in part representative of their protein products) that function in units, such as co-expressed or co-regulated genes, may not be uncovered. These gene units, or modules, have been hypothesized to characterize complex disease processes accurately, and importantly, they do so in the context of functionally related genes (Chaussabel and Baldwin 2014; Saelens et al. 2018). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Methods \n",
    "##### Discovery Cohort\n",
    "Here I describe the identification of mortality-related gene sets/signatures from the gene expression profiles of severely-ill ICU patients. The ICU patients were recruited from a hospital in Toronto, Canada, within the first day of admission (except for two patients recruited from the hospital ward) (Table 3.1). These patients exhibited severe respiratory failure and/or suspected pulmonary sepsis.  Patients were excluded if death was impending (within 12 hours), if blood collection was unattainable, or consent was withheld. Of these, 27 (32.9%) were confirmed to be infected with SARS-CoV-2 by subsequent viral PCR. Furthermore, this cohort displayed severe symptomatology (7.2 ± 0.55 SOFA scores 24H post sampling) and high mortality (24.4% [20/82]). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Load required packages \n",
    "library(tidyverse)\n",
    "library(magrittr)\n",
    "library(fastICA) # Package with the fastICA algorithm \n",
    "library(functionjunction) # My package for data analysis \n",
    "library(enrichR)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "TRUE"
      ],
      "text/latex": [
       "TRUE"
      ],
      "text/markdown": [
       "TRUE"
      ],
      "text/plain": [
       "[1] TRUE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read in data \n",
    "icu_dat <- read_rds(\"../create_tr_te/tr_te_dat.RDS\")[[\"icu\"]]\n",
    "expr  <- icu_dat$expr\n",
    "meta <- icu_dat$meta\n",
    "all(colnames(expr)[-1] == meta$sample_identifier)\n",
    "\n",
    "# Read in gene names\n",
    "gene_names <- read_rds(\"../../../sepsis_rnaseq_all/final/counts_meta_10_read_filt_261120.RDS\")$universe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>12238</li><li>83</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 12238\n",
       "\\item 83\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 12238\n",
       "2. 83\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 12238    83"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>ensembl_gene_id</th><th scope=col>sepcv001T1</th><th scope=col>sepcv002T0</th><th scope=col>sepcv003T0</th><th scope=col>sepcv004T0</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>ENSG00000000419</td><td>  706</td><td>  216</td><td>  384</td><td>  495</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>ENSG00000000457</td><td>  621</td><td>  181</td><td>  304</td><td>  504</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>ENSG00000000460</td><td>  100</td><td>   18</td><td>   74</td><td>   78</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>ENSG00000000938</td><td>14387</td><td>10431</td><td>18687</td><td>21591</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>ENSG00000000971</td><td>   61</td><td>   15</td><td>   11</td><td>   16</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & ensembl\\_gene\\_id & sepcv001T1 & sepcv002T0 & sepcv003T0 & sepcv004T0\\\\\n",
       "  & <chr> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t1 & ENSG00000000419 &   706 &   216 &   384 &   495\\\\\n",
       "\t2 & ENSG00000000457 &   621 &   181 &   304 &   504\\\\\n",
       "\t3 & ENSG00000000460 &   100 &    18 &    74 &    78\\\\\n",
       "\t4 & ENSG00000000938 & 14387 & 10431 & 18687 & 21591\\\\\n",
       "\t5 & ENSG00000000971 &    61 &    15 &    11 &    16\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 × 5\n",
       "\n",
       "| <!--/--> | ensembl_gene_id &lt;chr&gt; | sepcv001T1 &lt;dbl&gt; | sepcv002T0 &lt;dbl&gt; | sepcv003T0 &lt;dbl&gt; | sepcv004T0 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | ENSG00000000419 |   706 |   216 |   384 |   495 |\n",
       "| 2 | ENSG00000000457 |   621 |   181 |   304 |   504 |\n",
       "| 3 | ENSG00000000460 |   100 |    18 |    74 |    78 |\n",
       "| 4 | ENSG00000000938 | 14387 | 10431 | 18687 | 21591 |\n",
       "| 5 | ENSG00000000971 |    61 |    15 |    11 |    16 |\n",
       "\n"
      ],
      "text/plain": [
       "  ensembl_gene_id sepcv001T1 sepcv002T0 sepcv003T0 sepcv004T0\n",
       "1 ENSG00000000419   706        216        384        495     \n",
       "2 ENSG00000000457   621        181        304        504     \n",
       "3 ENSG00000000460   100         18         74         78     \n",
       "4 ENSG00000000938 14387      10431      18687      21591     \n",
       "5 ENSG00000000971    61         15         11         16     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "survive    dead \n",
       "     60      20 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Let get some basic info  \n",
    "dim(expr)\n",
    "functionjunction::first_five(expr)\n",
    "table(meta$mortality) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# RNA Seq data must be normalized prior to analysis due to heteroskedasticity of count data (variance depending on mean count)\n",
    "# My go to is the variance stabilizing transformation by DESeq2. Other options exist, like TPM, FPKM, RPKM, rlog, etc. \n",
    "expr_vst <- expr %>% \n",
    "  column_to_rownames(var = \"ensembl_gene_id\") %>% \n",
    "  as.matrix() %>% \n",
    "  DESeq2::varianceStabilizingTransformation() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### ICA \n",
    "\n",
    "There are several statistical techniques to identify modules from a gene expression matrix, which mostly rely on unsupervised learning techniques such as correlation networks (e.g., Weighted Gene Correlation Network Analysis/ WGCNA), clustering (e.g., Hierarchical, K-means), and matrix decomposition (e.g., PCA, Independent Component Analysis/ICA). Prior benchmarking studies concluded that ICA-derived modules outperformed those obtained by other methods when comparing inferred modules to known regulatory modules in E. coli, S. cerevisiae, and humans (Rotival et al. 2011; Saelens et al. 2018). \n",
    "\n",
    "In the context of gene expression analysis, ICA aims to uncover underlying biological processes (e.g., mechanisms mediating transcriptional regulation, signaling cascades, immune responses, etc.) represented by gene modules that yield the observed gene expression events. In other words, the observed gene expression matrix can be considered a net sum of unobserved or latent biological processes, which is inferred by the ICA matrix decomposition (briefly explained below). There are two notable advantages to this technique. Firstly, ICA-derived modules enable genes to participate in more than one module, unlike clustering and correlation network approaches which divide genes into just one cluster/module based on a set distance cut-off, which is biologically implausible. Secondly, the ICA-derived modules are statistically independent by construction (or as independent as possible), which allows the components to map to distinct biological processes influencing gene expression. These components are reminiscent of principal components obtained using the related method PCA, although PCA components have no requirement to be statistically independent and thus fail to map to independent biological processes. \n",
    "\n",
    "$$\n",
    "X_{ij} = \\sum_{k=1}^{K} S_{ik} A_{kj} + \\epsilon_{ij}\n",
    "$$\n",
    "\n",
    "\n",
    "I applied a pipeline previously used in the studies Teschendorff et al. (2007) and Rotival et al. (2011), which employs the FastICA algorithm to identify components (Hyvärinen and Oja 2000). Briefly, the input to the FastICA algorithm is a p × n gene expression matrix (where p represents genes and n represents patients/observations). A p × k “Signal” matrix is among the decomposed output matrix products. Here, the k columns represent a set of statistically-independent “components” (or random variables) describing the activation (or contribution) of the individual p genes in the various components. This approach was used to identify 37 components representing independent biological processes. The ICA components were reduced to a set of module genes that strongly influenced the component distribution (i.e., the genes at the extremes of distribution), as assessed by false discovery rate (FDR) estimates (Strimmer 2008; FDR<10-3). An “eigenmodule”, which represented the first principal component of the module gene expression matrix, was derived to summarize module gene expression into a single variable. This approach was inspired by the WGCNA method, which upon identifying highly correlated sets of gene modules summarizes module gene expression in a fashion identical to the one implemented (Langfelder and Horvath 2008). A multiple linear regression was performed to estimate the association of eigenmodules to eventual mortality, wherein the response variable was the eigenmodule, and covariates included binarized mortality status, age, sex, and transformed cell proportions (using PCA). Cell proportions were included in the linear model in case modules captured gene sets comprised of correlated gene expression of a specific cell type. Functional characterization of DE and module genes was performed using an overrepresentation analysis of Reactome and MSigDB Hallmarks (adjusted p-value ≤0.05) (Fabregat et al. 2018; Liberzon et al. 2015). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write the function to perform ICA \n",
    "perform_ica <- function(expr_mat, kurtosis_keep = 10, qval_filt = 0.001, seed = 1) {\n",
    "  \n",
    "  # Determine the number of components to learn using PCA\n",
    "  # Specifically, how many PCs explain 90% of variance in the expression matrix?  \n",
    "  comp_no <- expr_mat %>% \n",
    "    t() %>% \n",
    "    as.data.frame() %>% \n",
    "    remove_zero_var_cols() %>% \n",
    "    perform_pca() %>% \n",
    "    pluck(\"pov\") %>% \n",
    "    explain_95(perc = 95)\n",
    "  cat(paste0(\"Performing ICA with \", comp_no, \" components.\\n\"))\n",
    "  \n",
    "  # Perform ICA\n",
    "  set.seed(seed)\n",
    "  ICA <- fastICA::fastICA(expr_mat, n.comp = comp_no)\n",
    "  cat(\"ICA Done.\\n\")\n",
    "  \n",
    "  # Get the S matrix and rename the components as Modules\n",
    "  ICA_S <- ICA$S %>% \n",
    "    as.data.frame() %>% \n",
    "    set_names(paste0(\"Mod_\",   str_pad(1:ncol(.), width = 3, side = \"left\", pad = \"0\")    ))\n",
    "\n",
    "  # Keep certain components based on kurtosis. Components with a high degree of kurtosis indicates the distribution is skewed (deviates from normality), suggesting  \n",
    "  comp_keep <- ICA_S %>% \n",
    "    map_dbl(~e1071::kurtosis(.x)) %>%  \n",
    "    keep(~.x >= kurtosis_keep) %>% \n",
    "    names()\n",
    "  ICA_S_filt <- ICA_S[,comp_keep]\n",
    "  \n",
    "  # Filter components of interest with FDR estimation\n",
    "  ICA_mods <- ICA_S_filt %>% \n",
    "    map(~fdrtool::fdrtool(.x, plot = FALSE, verbose = FALSE))\n",
    "  \n",
    "  # Filter genes in each componennt \n",
    "  ICA_mods_filt <- list()\n",
    "  for (comp in names(ICA_mods)) {\n",
    "    mod_df <- data.frame(qval = ICA_mods[[comp]][[\"qval\"]], ensembl_gene_id = rownames(ICA_S))\n",
    "    mod_df <- mod_df %>% \n",
    "      dplyr::filter(qval <= qval_filt) %>% \n",
    "      left_join(gene_names, by = \"ensembl_gene_id\")\n",
    "    ICA_mods_filt[[comp]] <- mod_df\n",
    "  }\n",
    "  \n",
    "  return(list(ICA_S_filt = ICA_S_filt, ICA_mods_filt = ICA_mods_filt))\n",
    "}\n",
    "\n",
    "# PCA \n",
    "perform_pca <- function(expr_mat){\n",
    "    # Perform PCA with prcomp\n",
    "  df <- expr_mat %>% \n",
    "    as.data.frame() %>% \n",
    "    prcomp(center = TRUE, scale. = TRUE)\n",
    "  \n",
    "  PoV <- round(df$sdev^2/sum(df$sdev^2), digits = 3)\n",
    "  x <- df$x %>% \n",
    "    as.data.frame() %>% \n",
    "    rownames_to_column(var = \"sample_identifier\") \n",
    "  rot <- df$rotation %>% \n",
    "    as.data.frame() %>% \n",
    "    rownames_to_column(var = \"ensembl_gene_id\")\n",
    "  \n",
    "  res <- list(x = x, rot = rot,  pov = PoV)\n",
    "  cat(paste0(\"PCA ---- DONE.\\n\"))\n",
    "  return(res)\n",
    "}\n",
    "\n",
    "# Quick helper function to extract the number of PCs explaining X percent of variation in the data \n",
    "explain_95 <- function(ev, perc){\n",
    "  for (ind in 1:length(ev)){\n",
    "    var_expl <- sum(ev[1:ind])\n",
    "    if(var_expl > perc/100){\n",
    "      return(ind)\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PCA ---- DONE.\n",
      "Performing ICA with 56 components.\n",
      "ICA Done.\n"
     ]
    }
   ],
   "source": [
    "# Now run the ICA pipeline \n",
    "ICA_res <- perform_ica(\n",
    "    expr_mat = expr_vst, \n",
    "    kurtosis_keep = 10, \n",
    "    qval_filt = 0.001, \n",
    "    seed = 1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now that I have modules, I need to determine whether the gene modules participate in established pathways/hallmarks using over-representation analysis. \n",
    "go_enrichment <- function(gene_list, p_val = 0.05, ID = \"hgnc_symbol\"){\n",
    "  \n",
    "  # Check if the gene list is empty\n",
    "  if(length(gene_list) == 0){\n",
    "    return(NULL)\n",
    "  }\n",
    "  \n",
    "  gene_list = na.omit(gene_list)\n",
    "  \n",
    "  # Set res to NULL\n",
    "  res <- NULL\n",
    "  \n",
    "  # Get results\n",
    "  tryCatch({\n",
    "    res <- enrichR::enrichr(genes = as.character(gene_list), \n",
    "                            databases = c(\"MSigDB_Hallmark_2020\") \n",
    "    )\n",
    "  }, error = function(e){cat(\"ERROR :\", conditionMessage(e), \"\\n\")} )\n",
    "  \n",
    "  if (is.null(res)){\n",
    "    return(NULL)\n",
    "  } else {\n",
    "    res <- res %>% map(~dplyr::filter(.x, Adjusted.P.value <= p_val))\n",
    "    return(res)\n",
    "  }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n",
      "Uploading data to Enrichr... Done.\n",
      "  Querying MSigDB_Hallmark_2020... Done.\n",
      "Parsing results... Done.\n"
     ]
    }
   ],
   "source": [
    "# Perform mSigDB enrichment\n",
    "sink()\n",
    "ICA_mods_filt_msigdb_enr <-  ICA_res$ICA_mods_filt %>%\n",
    "    map(~pull(.x, hgnc_symbol)) %>%\n",
    "    map(~go_enrichment(.x,  p_val = 0.05))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_mod_eigenvect <- function(ICA_mods, expr_mat) {\n",
    "  \n",
    "  eigenmodule <- ICA_mods$ICA_mods_filt %>%\n",
    "    purrr::discard(~nrow(.x) == 0) %>% \n",
    "    map(~as.matrix(expr_mat)[.x$ensembl_gene_id, ]) %>%\n",
    "    map(~t(.x) %>% perform_pca() %>% pluck(\"x\")) %>%\n",
    "    map(~dplyr::select(.x, one_of(\"sample_identifier\", \"PC1\")))\n",
    "  \n",
    "  return(eigenmodule)\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Get Eigenvectors\n",
    "ICA_mod_eigen <- get_mod_eigenvect(ICA_res, expr_vst)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Associate modules to clinical vars_of_int\n",
    "lm_clin_eig <- function(ICA_eig, meta, vars_of_int, deconv = NULL, incl_cell_props = FALSE) {\n",
    "  met <- meta\n",
    "  \n",
    "  # Check things are in the right order \n",
    "  right_order <- ICA_eig %>% \n",
    "    map(~all(.x$sample_identifier == met$sample_identifier)) %>% \n",
    "    unlist() %>% \n",
    "    all()\n",
    "  \n",
    "  if (isFALSE(right_order)) {\n",
    "    return(c(\"Things are not in the right order\"))\n",
    "  }\n",
    "  \n",
    "  # Design formula \n",
    "  des = \"PC1 ~ \"\n",
    "  \n",
    "  # Include cell proportions  \n",
    "  if (incl_cell_props) { \n",
    "    \n",
    "    if (is.null(deconv)) {\n",
    "      stop()\n",
    "    }\n",
    "    \n",
    "    deconv_pca <- deconv %>% \n",
    "      rownames_to_column(var = \"sample_identifier\") %>% \n",
    "      filter(sample_identifier %in% meta$sample_identifier) %>% \n",
    "      remove_zero_var_cols() %>% \n",
    "      column_to_rownames(var = \"sample_identifier\") %>% \n",
    "      perform_pca() \n",
    "    \n",
    "    pcs_to_incl <- deconv_pca$pov %>% explain_95(perc = 90)\n",
    "    cat(paste0(\"Cell Proportion PCs used: \", pcs_to_incl, \"\\n\"))\n",
    "    pcs_to_incl <- paste0(\"PC\", 1:pcs_to_incl)\n",
    "    \n",
    "    deconv_pca_x <- deconv_pca$x %>% \n",
    "      dplyr::select(one_of(\"sample_identifier\", pcs_to_incl)) %>% \n",
    "      dplyr::rename_at(vars(matches(\"PC\")), ~ paste0(\"CiSo_\", .))\n",
    "    \n",
    "    met <- met %>% \n",
    "      left_join(deconv_pca_x,  by = \"sample_identifier\")\n",
    "    \n",
    "    if (length(unique(met$sequencing_month_year)) > 1 ) {\n",
    "      paste0(des, \" sequencing_month_year + \") # Need to fix des = paste0() ...\n",
    "    }\n",
    "    \n",
    "    des <- paste0(des, \"gender + age + \", paste(paste0(\"CiSo_\",pcs_to_incl), collapse = \" + \"), \" + \")\n",
    "  }\n",
    "  \n",
    "  # Initiate loop - loop through variables, and then module\n",
    "  res <- list()\n",
    "  for (var in  vars_of_int) {\n",
    "    for (mod in names(ICA_eig)) {\n",
    "      # Create a new df with the var and eigenvector the module\n",
    "      lm_df <- ICA_eig[[mod]] %>% \n",
    "        left_join(met, by = \"sample_identifier\")\n",
    "      \n",
    "      # GLM - PC1 ~ var\n",
    "      res[[var]][[mod]] <- glm(formula = formula(paste0(des,  var)), data = lm_df ) %>% \n",
    "        broom::tidy()\n",
    "    }\n",
    "  }\n",
    "\n",
    "  return(list(res = res))\n",
    "} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deconv_res <- read_rds(\"./create_tr_te/deconv_res.rds\") %>% \n",
    "  map(~map(.x, ~as.data.frame(.x)))\n",
    "deconv_res <- deconv_res$icu$cibersort\n",
    "\n",
    "ICA_mod_eigen_clin <- lm_clin_eig(ICA_mod_eigen, meta, vars, deconv_res, TRUE) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get MsigDB hallmarks for each module\n",
    "msigdb = ICA_mod_plot$expr_uncorr$msigdb[ICA_mod_expr_dir$comp]\n",
    "\n",
    "ICA_mods_filt_msigdb_enr_df <- msigdb %>%\n",
    "  map(~keep(.x, names(.) %in% c(\"MSigDB_Hallmark_2020\"))) %>%\n",
    "  map(~bind_rows(.x, .id = \"database\")) %>%\n",
    "  bind_rows(.id = \"comp\") %>% \n",
    "  separate(Overlap,into = c(\"M\", \"N\")) %>%\n",
    "  mutate(Ratio = as.numeric(M)/as.numeric(N)*100 ) %>%\n",
    "  mutate(Term = case_when(Term == \"heme Metabolism\" ~ \"Heme Metabolism\", TRUE ~ Term))  %>% \n",
    "  group_by(comp) %>%\n",
    "  top_n(2, Ratio) %>%\n",
    "  ungroup() %>% \n",
    "  dplyr::select(one_of(\"comp\", \"Term\", \"Ratio\")) %>% \n",
    "  right_join(data.frame(comp = names(msigdb))) %>% \n",
    "  #right_join(data.frame(comp = ICA_mod_expr_dir$comp  )) %>% \n",
    "  mutate(Ratio = ifelse(is.na(Ratio), 0, Ratio)) %>% \n",
    "  mutate(Term = ifelse(is.na(Term), \"None\", Term)) %>% \n",
    "  pivot_wider(id_cols = \"Term\", names_from = \"comp\", values_from = \"Ratio\") %>% \n",
    "  na2zero() %>% \n",
    "  filter(!Term == \"None\") %>% \n",
    "  mutate(database = \"mSigDB\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the clinical association between each module and an outcome   \n",
    "ICA_mod_eigen_clin_df <- ICA_mod_eigen_clin$expr_uncorr$res %>%\n",
    "  map(~bind_rows(.x, .id = \"comp\")) %>%\n",
    "  bind_rows(.id = \"variable\") %>%\n",
    "  filter(term != \"(Intercept)\" ) %>% \n",
    "  mutate(adj_p_value = p.adjust(p.value, method = \"BH\")) %>% \n",
    "  filter(str_detect(term,  paste(vars, collapse = \"|\") )) %>% \n",
    "  mutate(p_val_bin = case_when(\n",
    "    adj_p_value >0.05 ~ \"P>0.05\", \n",
    "    adj_p_value >0.01 ~ \"P>0.01\",\n",
    "    adj_p_value >0.001 ~ \"P>0.001\",\n",
    "    adj_p_value >0.0001 ~ \"P>0.0001\", \n",
    "    TRUE ~ \"P<0.0001\")\n",
    "    ) %>% \n",
    "  mutate(log10_p.value = -1*log10(p.value)) %>% \n",
    "  dplyr::select(one_of(\"variable\", \"comp\", \"log10_p.value\", \"p_val_bin\", \"p.value\")) %>% \n",
    "  pivot_wider(id_cols = \"comp\",names_from = \"variable\", values_from = \"p_val_bin\") %>% \n",
    "  #arrange(match(comp, colnames(ICA_mods_filt_df)[-c(1:2)] ))\n",
    "  arrange(match(comp, ICA_mod_expr_dir$comp))\n",
    "\n",
    "#### Start making the heatmap\n",
    "all(colnames(ICA_mods_filt_df)[-c(1:2)] == ICA_mod_eigen_clin_df$comp)\n",
    "colnames(ICA_mods_filt_df) <- str_replace(colnames(ICA_mods_filt_df), \"_\", \" \")\n",
    "\n",
    "library(ComplexHeatmap)\n",
    "#col_fun <- circlize::colorRamp2(c(0,1), c( \"white\", \"#330033\"))\n",
    "col_fun = structure(c( \"gray95\", \"#330033\"), names = c( \"Not Enr.\", \"Enr. Hallmark\"))\n",
    "col_assoc = c(\"P>0.05\" = \"grey60\", \n",
    "              \"P>0.01\" = \"darkgoldenrod1\",\n",
    "              \"P>0.001\"= \"darkmagenta\",\n",
    "              \"P>0.0001\"= \"mediumblue\",\n",
    "              \"P<0.0001\"= \"midnightblue\"\n",
    "              ) \n",
    "col_dir = c(\"Up\" = \"firebrick\" ,\"Down\"=\"darkgreen\", \"Mixed\" = \"blue\")\n",
    "column_ha <- HeatmapAnnotation(\n",
    "  #Culture = ICA_mod_eigen_clin_df$culture,\n",
    "  #\"Organ Dysf. Group\" = ICA_mod_eigen_clin_df$sofa_24,\n",
    "  \"Eigenmod/Mortality Assoc.\" = ICA_mod_eigen_clin_df$mortality,\n",
    "  \"Module Expr Direction\" = ICA_mod_expr_dir$Direction,\n",
    "  col = list(\n",
    "    #Culture = col_assoc,\n",
    "    #\"Organ Dysf. Group\" = col_assoc,\n",
    "    \"Eigenmod/Mortality Assoc.\" = col_assoc,\n",
    "    \"Module Expr Direction\" = col_dir),\n",
    "  annotation_legend_param = list(\n",
    "    # Culture = list(\n",
    "    #   title = \"Eigenmod/\\nEndpoint Assoc. \\n-log10(AdjPValue)\"),\n",
    "    #\"Organ Dysf. Group\" = list(\n",
    "    #  title = \"Eigenmod/\\nEndpoint Assoc. \\n-log10(AdjPValue)\"),\n",
    "    \"Eigenmod/Mortality Assoc.\" = list(\n",
    "      title = \"Eigenmod/\\nEndpoint Assoc. \\n-log10(AdjPValue)\"\n",
    "      #labels = c(\"P=1\", \"P=0.05\", \"P=0.01\",\"P=0.001\",\"P=0.0001\" )\n",
    "      #at = c(0,1.3,2,3,4)\n",
    "      ),\n",
    "    \"Module Expr Direction\" = list(\n",
    "      title = \"Module Expr Direction\"\n",
    "    )\n",
    "  ),\n",
    "  border = TRUE,\n",
    "  #annotation_width=unit(c(2, 5.0), \"cm\"),\n",
    "  simple_anno_size = unit(1, \"cm\"), \n",
    "  gap = unit(1, 'mm'),\n",
    "  show_legend = c(TRUE, TRUE, TRUE)\n",
    "  )\n",
    "\n",
    "htmap <- ICA_mods_filt_df %>% \n",
    "  filter(database == \"mSigDB\") %>% \n",
    "  dplyr::select(-database) %>% \n",
    "  column_to_rownames(var = \"Term\") %>% \n",
    "  as.matrix() %>% \n",
    "  Heatmap(\n",
    "    name = \"\\nRatio\",\n",
    "    column_split = ICA_mod_expr_dir$Direction,\n",
    "    cluster_rows = FALSE,\n",
    "    cluster_columns = FALSE,\n",
    "    top_annotation = column_ha,\n",
    "    col = col_fun,\n",
    "    border = TRUE\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
